### PR TITLE
New features for dgroc

### PR DESCRIPTION
**Pull request by vojtechtrefny**
I started using dgroc to daily build some of my projects and found few problems/missing features. I think my changes could be useful for others too.

Changes:
- python-copr support. It's easier to use (from python) than the API and it could upload the SRPMs to copr directly.
- Custom command for creating source archive. Just creating an archive from git (mercurial) is often not enough -- it might be necessary to run some autogen/configure magic etc.
- Updating spec files from upstream repository. If the project is already packaged in Fedora and it has its spec file in the repository, it might be useful to update "our" spec file when there is a version change -- so the version (+ build number) in copr is always greater than version available from Fedora repos.

### CODE REVIEW

**Comment by vojtechtrefny** [comment]
Updated version -- I've removed the third commit (updating spec files from upstream repository) -- I need to come up with new "more clever" solution for this problem.

File: dgroc.py [hunk 1]
```diff
@@ -76,8 +78,7 @@ def commit_hash(cls, folder):
     @classmethod
     def archive_cmd(cls, project, archive_name):
         '''Command to generate the archive'''
-        return ["git", "archive", "--format=tar", "--prefix=%s/" % project,
-           "-o%s/%s" % (get_rpm_sourcedir(), archive_name), "HEAD"]
+        return "git archive --format=tar --prefix=%s/ -o%s/%s HEAD" % (project, get_rpm_sourcedir(), archive_name)
```
Comment (pypingou) [inline dgroc.py]: Could you split these long lines over multiple row?

**Comment by pypingou** [comment]
This PR sound good but is rather large, so I'll have to go through it carefully.

**Comment by vojtechtrefny** [comment]
> This PR sound good but is rather large, so I'll have to go through it carefully.

Ok. If you have any questions, feel free to ask here or on IRC (vtrefny at FreeNode).

File: dgroc.py [hunk 2]
```diff
@@ -108,13 +109,11 @@ def commit_hash(cls, folder):
     @classmethod
     def archive_cmd(cls, project, archive_name):
         '''Command to generate the archive'''
-        return ["hg", "archive", "--type=tar", "--prefix=%s/" % project,
-           "%s/%s" % (get_rpm_sourcedir(), archive_name)]
+        return "hg archive --type=tar --prefix=%s/ %s/%s" % (project, get_rpm_sourcedir(), archive_name)
```
Comment (pypingou) [inline dgroc.py]: Same as above about the length

File: dgroc.py [hunk 3]
```diff
@@ -229,6 +215,61 @@ def get_rpm_sourcedir():
...
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+        shell=True)
+    out = pull.communicate()
+
+    # no archive name -> need to read it from output and move the archive to rpm source dir
+    if not archive_name:
+        archive_name = _get_archive_name(out, project)
+        if not os.path.isfile(git_folder + "/" + archive_name):
+            LOG.info('Got archive name %s, but no such file found in %s' % (archive_name, git_folder))
```
Comment (pypingou) [inline dgroc.py]: And length here as well :)

**Comment by pypingou** [comment]
The ``requirements.txt`` file should also be updated btw :)

**Comment by pypingou** [comment]
I tried to make it run but I ran into:
````
SRPM built: /home/pingou/rpmbuild/SRPMS/guake-0.7.2-28.20151006gitb500576e.el6.src.rpm
1 srpms generated
Reading configuration for copr
Server error, raw response body: {
  "error": "Invalid request", 
  "output": "notok"
}
Something went wrong:
  Invalid request
Monitoring 0 builds...
````
So no idea what's going on here, do you?

**Comment by pypingou** [comment]
Any idea?

**Comment by vojtechtrefny** [comment]
Ok, I fixed the long line.

The error -- I'm not sure, "Server error" should be the equivalent of server returning 5xx message. Maybe, are you sure your copr API token is valid (it expires after 6 months or so)? My token expired a month ago and I got really weird error because of that (but I don't remember what the error message said).

**Comment by pypingou** [comment]
I also thought about the API key, but it says ``# expiration date: 2015-12-04`` :-/

But I think this reveals a flow in the current software, we need the possibility to return more debug information to the user :)

**Comment by vojtechtrefny** [comment]
I've added more logs and also the logger from copr.client, hope it will help.

I also found a "bug" in my code -- I'm totally ignoring `copr_url` set in `.config/dgroc`, I'm just reading everything from `.config/copr`. But this shouldn't cause this problem (because the error occurs when creating new build, not when creating the copr client instance). I will fix this ASAP.

**Comment by vojtechtrefny** [comment]
Took me little longer (too much programming at work and not enough free time) but reading `copr_url` from `.config/dgroc` is back.

**Comment by azrdev** [comment]
Hey, it would be great if dgroc could upload its SRPM directly via `copr-cli` -- looks like this is part of c5b7eb3, what's keeping that one from being merged?

**Pull request closed without merging**